JBPM-5808: Elements are placed in preview position

Christopher-Chianelli opened the pull request:
Elements are now centered on the cursor when placed.

manstis commented:
@Christopher-Chianelli Hi.. I assume this fix is to ensure when a new {{Shape}}  is dragged from the palette it's canvas coordinate is such that it appears centred around the mouse pointer position on drag-end? If so, does this change not also interfere with {{NodeBuilderControlImpl}} and hence the target coordinate for drag-end operations should be derived in {{ObserverBuilderControl}}? This would mean moving the calculation of the target location to a protected method in {{AbstractElementBuilderControl}} and override it in {{ObserverBuilderControl}} to your change?

Also, a unit test would be nice.

Christopher-Chianelli commented:
Added unit tests @manstis. I don't really understand what you mean by "If so, does this change not also interfere with {{NodeBuilderControlImpl}} and hence the target coordinate for drag-end operations should be derived in {{ObserverBuilderControl}}? This would mean moving the calculation of the target location to a protected method in {{AbstractElementBuilderControl}} and override it in {{ObserverBuilderControl}} to your change?". The only change is in {{AbstractElementBuilderControl}}, inside getElementCommands (which returns the commands needed to create the node and move it in position). It does not appear this change would affect {{NodeBuilderControlImpl}}, since {{NodeBuilderControlImpl}} does not call getElementCommands either directly or indirectly.

manstis commented:
@Christopher-Chianelli Thanks for the test.

Please see [here](https://github.com/kiegroup/kie-wb-common/blob/master/kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/builder/impl/NodeBuilderControlImpl.java#L124). This gets a ```ElementBuilderControlImpl``` that extends ```AbstractElementBuilderControl```. 

This [line](https://github.com/kiegroup/kie-wb-common/blob/master/kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/builder/impl/NodeBuilderControlImpl.java#L131) then calls ```getElementCommands(...)``` that calls your new code to position the node at ```(requestedX-shapeWidth/2, requestedY-shapeHeight/2)```. 

This is great when using the ```ObserverBuilderControl``` sub-class as you want the "drop" position to be centred around the mouse pointer. But with ```NodeBuilderControlImpl``` if I have programmatically requested a Shape to be at ```(x, y)``` I expect it to be placed at ```(x, y)``` and not ```(x-width/2, y-width/2)```.

Am I wrong?

Christopher-Chianelli commented:
Ah, I see what you mean. I thought `AbstractElementBuilderControl` was only used by UI. Thanks for telling me the UI class; I added a protected method to get the target location, defaults to the original to-left position, changed to shape center in `ObserverBuilderControl`.

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/test/java/org/kie/workbench/common/stunner/client/lienzo/canvas/controls/builder/ObserverBuilderControlTest.java:
```diff
@@ -0,0 +1,274 @@
+        when(shapeManager.getShapeSet(anyString())).thenReturn(shapeSet);
+        when(shapeManager.getDefaultShapeSet(anyString())).thenReturn(shapeSet);
+
+        tested = new ObserverBuilderControl(clientDefinitionManager,
```
Formatting :-(

Christopher-Chianelli replied:
Odd, despite having it set to reformat my code on commit, it appears the code isn't being reformatted. Time to file a bug report to IntelliJ...

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/test/java/org/kie/workbench/common/stunner/client/lienzo/canvas/controls/builder/ObserverBuilderControlTest.java:
```diff
@@ -0,0 +1,274 @@
+
+        tested.enable(canvasHandler);
+        tested.setCommandManagerProvider(commandManagerProvider);
+    }
```
Ah.. the sweet smell of mocking the hell out of something :-)

manstis commented:
@romartin want to review too?

Christopher-Chianelli commented:
Reformatted code.

hasys commented:
Jenkins execute full downstream build

romartin commented:
Hey @Christopher-Chianelli @hasys 
See comments in the ticket, I would vote, as from feedback's from UX, to change this as:
- Use top-left corner for the icon (while dragging from the palette)
- Use top-left corner for the resulting shape (once dropping)
- If possible, while dragging from the palette, change the cursor to pointer
Thanks!

hasys commented:
Hi @romartin,

yes, sounds great, let's synchronize this issues with UX guys comments.

Christopher-Chianelli commented:
@romartin, I have integrated your suggested changes; however, the cursor's icon is flickery.

romartin commented:
yep @Christopher-Chianelli , please see comments in the jira ticket, let's try to fix this! :) Thanks

hasys commented:
Hi @Christopher-Chianelli,

Thank you, just want to mention that ObserverBuilderControlTest#testAddElement test fails locally as well.

hasys commented:
Jenkins execute full downstream build

Christopher-Chianelli commented:
Hey @romartin @hasys,

I fixed the test (which was testing to see if the shape was dropping centered on the mouse cursor, which is changed) and fixed the icon flickering.

hasys commented:
Jenkins execute full downstream build

hasys commented:
Jenkins retest please

romartin commented:
Hey @Christopher-Chianelli , tested and all good, just added a couple of minor things and ready to go! 
thanks!! :)

manstis commented:
Jenkins please retest this.

Christopher-Chianelli commented:
Jenkins please retest this

manstis commented:
@Christopher-Chianelli You really need to see some of these ```kie-wb-common``` PRs through to completion!

Christopher-Chianelli commented:
@manstis This one is finished, but is failing due to bad wildfly-dist property. Let me know if/when the PR for updating that property is merged.

Christopher-Chianelli commented:
Jenkins execute full downstream build

Christopher-Chianelli commented:
Note: the test that failed was `testIndexEnumEntries`, which, if I recall correctly, can be a false negative due to how the test was made and how Jenkins run threads

Christopher-Chianelli commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/builder/impl/AbstractElementBuilderControl.java:
```diff
@@ -258,16 +258,22 @@ public void getElementCommands(final Element element,
 
+    protected Point2D getTargetCords(Node<View<?>,Edge> node, double x, double y) {
+        return new Point2D(x,y);
+    }
```
@hasys, that is an artifact from a previous version. That method is meant to be overridden by subclasses if they want to change where the shape is placed. Previously, I made it so shapes placed centered on the mouse cursor in `ObserverBuilderControl`. In order to do that, `ObserverBuilderControl` needed to know the shape size, thus the node parameter. It was later decided to place shapes from the top-left corner, so I undo my changes in `ObserverBuilderControl` and updated `ShapeGlyphDragHandlerImpl` to be based from the top-left of the mouse cursor instead of centered. The method can be removed since `ObserverBuilderControl` doesn't override it anymore.

Christopher-Chianelli commented:
@romartin @hasys I removed the pointless method and removed the unused imports.

romartin commented:
All good, ready to merge for me! !
CC @manstis 
Thanks

manstis commented:
@hasys Good for you?

romartin commented:
Jenkins please retest this

romartin commented:
NOTE: Just fired a rebuild to ensure this still works after @hasys code style refactoring....

hasys commented:
Jenkins execute full downstream build

romartin commented:
@hasys the idea is the icons that appear during drag&drop are the same ones that should appear in the palette and other components, like the diagram explorer. I mean if some icon is showing shadows is not due to it's being dragged... it's because the icon contains shadows. Not sure if I'm following you yet @hasys hehe... what you mean?
